llama-index-integrations/llms/llama-index-llms-nvidia/llama_index/llms/nvidia/base.py

```diff
@@ -0,0 +1,379 @@
+        self,
+        model: str = DEFAULT_CATALOG_MODEL,
+        temperature: float = DEFAULT_TEMPERATURE,
+        max_tokens: int = DEFAULT_CATALOG_MAX_TOKENS,
+        timeout: float = 120,
+        max_retries: int = 5,
+        nvidia_api_key: Optional[str] = None,
+        api_key: Optional[str] = None,
+        callback_manager: Optional[CallbackManager] = None,
+        system_prompt: Optional[str] = None,
+        messages_to_prompt: Optional[Callable[[Sequence[ChatMessage]], str]] = None,
+        complettion_to_prompt: Optional[Callable[[str], str]] = None,
+        pydantic_program_mode: PydanticProgramMode = PydanticProgramMode.DEFAULT,
+        output_parser: Optional[BaseOutputParser] = None,
+    ) -> None:
+        callback_manager = callback_manager or CallbackManager([])
+
+        api_key = get_from_param_or_env(
+            "api_key",
+            nvidia_api_key or api_key,
+            "NVIDIA_API_KEY",
+            "NO_API_KEY_PROVIDED",
+        )
+
+        self._client = SyncOpenAI(
+            api_key=api_key,
+            base_url=BASE_CATALOG_URL,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        self._client._custom_headers = {"User-Agent": "llama-index-llms-nvidia"}
+        self._aclient = AsyncOpenAI(
+            api_key=api_key,
+            base_url=BASE_CATALOG_URL,
+            timeout=timeout,
+            max_retries=max_retries,
+        )
+        self._aclient._custom_headers = {"User-Agent": "llama-index-llms-nvidia"}
+
+        super().__init__(
+            model=model,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            timeout=timeout,
+            max_retries=max_retries,
+            callback_manager=callback_manager,
+            system_prompt=system_prompt,
+            messages_to_prompt=messages_to_prompt,
+            complettion_to_prompt=complettion_to_prompt,
+            pydantic_program_mode=pydantic_program_mode,
+            output_parser=output_parser,
+        )
+
+    @property
+    def available_models(self) -> List[Model]:
+        ids = API_CATALOG_MODELS.keys()
+        if self._mode == "nim":
+            ids = [model.id for model in self._client.models.list()]
+        return [Model(id=name) for name in ids]
+
+    @classmethod
+    def class_name(cls) -> str:
+        return "NVIDIA"
+
+    @property
+    def metadata(self) -> LLMMetadata:
+        params = {
+            "num_output": self.max_tokens,
+            "is_chat_model": True,
+            "model_name": self.model,
+        }
+        if context_window := catalog_modelname_to_contextsize(self.model):
+            params["context_window"] = context_window
+        return LLMMetadata(**params)
+
+    @property
+    def _model_kwargs(self) -> Dict[str, Any]:
+        return {
+            "model": self.model,
+            "temperature": self.temperature,
+            "max_tokens": self.max_tokens,
+        }
+
+    def mode(
```

Comment:
Is it worth having this method? Wouldn't the user just create a new `NVIDIA()` llm object instead of switching a mode?

Reply:
right now this is a design point for the class. i'd like to include this now and later.